singer/src/main/java/com/pinterest/singer/reader/thrift/MappedFileTBinaryProtocol.java

```diff
@@ -0,0 +1,405 @@
+/**
+ * Copyright 2019 Pinterest, Inc.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.pinterest.singer.reader.thrift;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.io.UnsupportedEncodingException;
+import java.nio.ByteBuffer;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileChannel.MapMode;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TField;
+import org.apache.thrift.protocol.TList;
+import org.apache.thrift.protocol.TMap;
+import org.apache.thrift.protocol.TMessage;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.protocol.TProtocolException;
+import org.apache.thrift.protocol.TSet;
+import org.apache.thrift.protocol.TStruct;
+import org.apache.thrift.protocol.TType;
+import org.apache.thrift.transport.TTransport;
+
+/**
+ * SingerTBinaryProtocol is an implementation of TBinaryProtocol reading data
+ * from Memory Mapped File. The entire length of file is memory mapped using
+ * {@link FileChannel}. This only creates a memory mapping however data will
+ * only be read when actual bytes are accessed so this is safe from a memory
+ * utilization. It's the responsibility of the underlying operating system to
+ * swap pages if sufficient memory is not available.
+ * 
+ * Note that {@link TTransport} can't be used in this case since the
+ * {@link TProtocol} needs direct reference to the ByteBuffer for lower level
+ * byte manipulations which won't be possible without direct reference.
+ */
+public class MappedFileTBinaryProtocol extends TProtocol {
```

Comment:
shall we extend from`TBinaryProtocol`?

Reply:
That wouldn't add value since all methods have to be overridden. It calls TTransport methods for byte reads under the hood which we are trying to avoid since it defeats the purpose of this change. This change does special reads for ByteBuffer / readBinary by performing a slice creation.